### PR TITLE
MAINT/ENH: Support for weights and range when estimating optimal number of bins

### PR DESCRIPTION
**Pull request by nayyarv**
Following on from PR #6029, the methods provided ignore range or weights so this PR aims to address this. 

The estimators are still defined as separate functions if there is ever a future need to expose these estimators to users. Each estimator now takes x and weights as arguments. These subfunctions expectation of `weights` is that it is count-like. If it's a probability, the subfunctions won't choose the datasize properly.

`_hist_optim_numbins_estimator` handles the range keyword by masking the appropriate parts of `a` and `weights`. It also checks whether weights is of an appropriate dtype (int or float). If weights is probability-like (i.e. `weights.sum()<a.size`), I scale weights such that `weights.sum()=a.size`. This means that I effectively choose  `n=max(a.size, weights.sum())` for my datasize, allowing for decent behaviour whether weights is count-like or probability like, while keeping things simple.

Freedman Diaconis requires a weighted IQR to be calculated, and since `np.percentile` or `np.partition` does not provide weighted options, this method is only used for unweighted data. Trying to use the 'FD' estimator with weights results in a `TypeError`. 'auto' now chooses between 'sturges' and 'scott' for weighted data, while retaining original behaviour for unweighted data. 

I've put some basic tests together and minor changes to the documentation and would love to hear others thoughts and feedback on the design choices/implementation/style etc.

### CODE REVIEW

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -91,16 +102,37 @@ def _hist_optim_numbins_estimator(a, estimator):
     if a.size == 0:
         return 1
 
-    def sturges(x):
+    # Filter the data so we stay in the required range
+    if data_range is not None:
+        mn, mx = data_range
+        keep = (a >= mn)
+        keep &= (a <= mx)
+        if not np.logical_and.reduce(keep):
```
Comment (shoyer) [inline numpy/lib/function_base.py]: use `np.all` instead

File: numpy/lib/function_base.py [hunk 2]
```diff
@@ -91,16 +102,37 @@ def _hist_optim_numbins_estimator(a, estimator):
...
+            a = a[keep]
+            if data_weights is not None:
+                data_weights = data_weights[keep]
+
+    if data_weights is not None:
+        if data_weights.dtype != int and data_weights.dtype != float:
```
Comment (shoyer) [inline numpy/lib/function_base.py]: probably better to use `np.issubdtype` instead to allow for, e.g., float32

File: numpy/lib/function_base.py [hunk 3]
```diff
@@ -91,16 +102,37 @@ def _hist_optim_numbins_estimator(a, estimator):
...
+            raise TypeError("Unsupported weight type {} passed into "
+                            "histogram automatic bin estimator".format(data_weights.dtype))
+        if data_weights.sum() < a.size:
+            # we assume that weights represents counts. In the case that weights
+            # represents a probability (i.e. weights.sum()<=1),
+            # we scale the weights to represent a count where weights.sum = a.size
+            data_weights = data_weights * (a.size/data_weights.sum())
```
Comment (shoyer) [inline numpy/lib/function_base.py]: Why not simply set `data_size = np.maximum(a.size, weights.sum())` directly? I don't think there is anything else which needs a rescaled `weights`

File: numpy/lib/function_base.py [hunk 4]
```diff
@@ -141,14 +186,14 @@ def fd(x):
...
-    def auto(x):
+    def auto(x, weights=None):
         """
         The FD estimator is usually the most robust method, but it tends to be too small
         for small X. The Sturges estimator is quite good for small (<1000) datasets and is
         the default in R.
         This method gives good off the shelf behaviour.
         """
-        return max(fd(x), sturges(x))
+        return max(scott(x, weights), sturges(x, weights)) if weights is not None else max(fd(x), sturges(x))
```
Comment (shoyer) [inline numpy/lib/function_base.py]: this is too much logic for one line

Comment (shoyer) [inline numpy/lib/function_base.py]: also, it's worth writing a comment with the reasoning for the switch

**Comment by shoyer** [comment]
The design choices here look good to me.

**Comment by seberg** [comment]
Frankly, I am a bit sceptical about automatically guessing the weight type. Maybe it is clear enough, just wanted to say that it has a bit of a bad taste to me for starters.

That said, I think the functionality is not bad at all.

File: numpy/lib/function_base.py (hunk 1 above)
Comment (nayyarv) [inline numpy/lib/function_base.py]: I coped the filter step from https://github.com/numpy/numpy/blob/master/numpy/lib/function_base.py#L400 which came up in #6100. 

It's actually faster than `np.all`

**Comment by nayyarv** [comment]
I agree that choosing data_size from weights is far from perfect, but without further input from the user (`**kwargs` or another argument in `histogram`)/exposing the estimators, it's the best I can do.
After our discussion at the end of the last PR, we decided the two main use cases for weights + automatic bin estimation were either count-like or probability-like weights (matched up to friends/colleagues thoughts)
If the weights were count-like, it's sum would be greater than x.size and if it was probability-like, it's sum would be less, which is why I use the `max(a.size, weights.sum())` logic. 
If the user was using the weights option as a bin summation method, it's unlikely they would use automatic estimators, and rather fixed bin edges - so situations in which the weights are not count/probability like can be ignored. 

Any better solutions, I'm all ears. If nothing else, we could simply patch in the `range` support and deal with weights at a later date.

**Comment by seberg** [comment]
I am not currently convinced that you can always correctly infer the type. So I am still wondering if we cannot find a way to force the user to make a conscies choice.

**Comment by nayyarv** [comment]
Fair enough - the way I see it, we need to know 2 things
1. Whether or not to consider the weights when estimating the number of bins
2. If we do consider the weights, what should the data size be - `weights.sum()` or `a.size`. 

Basically, we need to know how to deal with datasize - if it's None, we can ignore the weights when estimating, if it's 'sum' or 'size', we know that we have to deal with weights. I.e. 1 argument.

Possible solutions,
1. Expose the estimators for manual control and let `np.histogram` provide simple default behaviour where it tries to guess (Maybe move the estimators to `scipy.stats`?)
2. Add another argument to `np.histogram` for instructions on how to deal with weights, `weights_sum=True` or something.
3. Add options to the `bins` string, like `'auto sum'` or `'scott size'`, which can then be `split` to retrieve the estimator and how to deal with weights. If the second term isn't included, ignore weights in the estimation. This also means the `bins` checking logic remains simple, as opposed to making it a tuple or list which is also iterable.
4. Something similar to 3, except maybe combined into the `weights` argument as a tuple or iterable?

I'm in favour of 3 (or maybe 1), as it requires the least amount of API changes, and it's an incremental change.

**Comment by njsmith** [comment]
Would it make sense to add a new kwarg data_size whose value defaults to
weights.sum()?

**Comment by nayyarv** [comment]
I'm not overly keen on adding a new kwarg that only applies when a bins estimator is being used and `weights` is not None. Seems like overkill, and I would like to avoid R style functions with their mostly unused arguments.

**Comment by njsmith** [comment]
I'm not keen on having lots of non-orthogonal kwargs either, but I like it
a lot better than the idea of defining a little ad hoc string language
inside a single string kwarg. The other general option that comes to mind
would be using something that has more structure than a str, like a dict or
the object, but I'm not seeing a terribly natural way to apply that trick
here myself...

**Comment by tacaswell** [comment]
If there are only two options for how to deal with the weights, maybe just do a look up on the whole string.  You effectively double/triple  the size of the number of estimators, but that does not seem so bad.  

From the mpl side, having `histogram` gain any kwargs would greatly complicate the argument handling in `hist`.

**Comment by shoyer** [comment]
I wonder if it would make sense to pick `data_size` based on the dtype of the array:

- int: use `data.sum()`
- float: use `data.size`

I guess this would interact poorly with downstream libraries like pandas that use floats to represent nullable-int.

**Comment by njsmith** [comment]
Is using data.size even theoretically justified? I can't quite think of any
reason why it's a more natural choice than any other arbitrary number
(though I admit my intuition about these algorithms is very limited).

On Wed, Sep 9, 2015 at 11:41 PM, Stephan Hoyer <notifications@github.com>
wrote:

> I wonder if it would make sense to pick data_size based on the dtype of
> the array:
>
>    - int: use data.sum()
>    - float: use data.size
>
> I guess this would interact poorly with downstream libraries like pandas
> that use floats to represent nullable-int.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6288#issuecomment-139131087>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

**Comment by nayyarv** [comment]
Haha, inferring `data_size` from the dtype is what we're unhappy with, so we're looking at a way to get the user to give us input without changing function signatures too much.

The estimator is primarily concerned with the size of the data to minimise Mean Integrated Squared Error. In general optimal binwidth ~ n^(-1/3) where n is the datasize.

Possible situations
1. Now if weights represents counts, this means that `np.repeat(a, weights)` would give us the full data set and hence n should be `weights.sum()`. This makes sense for integer weights and some float weights (if there's reasoning behind it).
2.In a situation where the weights represents a probability/proportion, the calculation of standard deviation or the IQR for scott's and FD's rule respectively are now altered, while the datasize should be the number of points, not `weight.sum()` which would evaluate to 1.
3. Finally, if there is no meaning to the weights, it's simply for the summation, the weights should be ignored for the calculation of the standard deviation and the datasize. This is for something exotic like complex weights or you have matched pairs eg. (Age, Wealth)  that you want to sum up per group.

We could potentially turn `bins` into a dict (or a list or tuple) for automatic methods, i.e,
`{method='auto', weights=['sum', 'size', 'ignore']}` for example, and for `bins='auto'` we either ignore weights by default or try and guess (`max(weights.sum(), a.size)` for data_size) for best average case behaviour.

**Comment by seberg** [comment]
Yeah, the whole discussion is because I am unhappy with guessing what to
do with weights based on the type. I have the feeling people are not
necessary enough aware of their type. And accuracy weights might add up
to more then the sample size in some cases as well....

**Comment by seberg** [comment]
I would like to throw in one further option, which may or may not be temporary:

Ignore the problem ;), maybe by exposing the bin number functions. Then just throw an error and tell user to use those functions manually. If weights are not very common (I have no clue) for plotting, we could do that reasonably.

**Comment by tacaswell** [comment]
I like @seberg's last suggestion (ignore/raise) the best.

**Comment by shoyer** [comment]
So just raising an error is weights are set? That seems reasonable to me, better than designing an awkward API for niche uses.

**Comment by njsmith** [comment]
It would be nice to support count weights + auto binning at some point, but
we can always do that as an enhancement later once we've figured out how to
actually make it work right :-). And until then, raising an error is safe
way to keep our options open, so +1

On Thu, Sep 10, 2015 at 11:08 AM, Stephan Hoyer <notifications@github.com>
wrote:

> So just raising an error is weights are set? That seems reasonable to me,
> better than designing an awkward API for niche uses.
>
> On Thu, Sep 10, 2015 at 10:18 AM, Thomas A Caswell
> <notifications@github.com> wrote:
>
> > I like @seberg's last suggestion (ignore/raise) the best.
> > On Thu, Sep 10, 2015 at 7:08 AM seberg <notifications@github.com> wrote:
> >> I would like to throw in one further option, which may or may not be
> >> temporary:
> >>
> >> Ignore the problem ;), maybe by exposing the bin number functions. Then
> >> just throw an error and tell user to use those functions manually. If
> >> weights are not very common (I have no clue) for plotting, we could do
> that
> >> reasonably.
> >>
> >> —
> >> Reply to this email directly or view it on GitHub
> >> <https://github.com/numpy/numpy/pull/6288#issuecomment-139204884>.
> >>
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/numpy/numpy/pull/6288#issuecomment-139315898
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6288#issuecomment-139330477>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

**Comment by nayyarv** [comment]
Ok, in that case, I'll edit the PR to support the `range` keyword only and throw an error for `weights not None`. R doesn't provide an option for `weights` with it's `hist` method (which is primarily visual), and I'm not sure how many situations with `'auto'` bin estimation will have weights too. I suppose we'll find out after 1.11 is released.

With regards to exposing the estimators, would it find a better home in `scipy.stats` instead?

**Comment by seberg** [comment]
It is true that exposing would fit better into scipy. And I am not a fan of exposing them in top level numpy. But scipy is not a dependency so that would be code duplication.
So, I know I suggested that we may expose it, but I frankly have no clue how it would look like :(.

**Comment by nayyarv** [comment]
Removed support for weights, and instead raise a TypeError when weights!=None is passed into the function. Support for range still remains, and the test is included too.

File: numpy/lib/function_base.py [hunk 5]
```diff
@@ -91,6 +101,17 @@ def _hist_optim_numbins_estimator(a, estimator):
     if a.size == 0:
         return 1
 
+    if data_weights is not None:
+        raise TypeError("Automated estimation of the number of "
```
Comment (nayyarv) [inline numpy/lib/function_base.py]: Maybe a ValueError? What's the best error to raise here?

**Comment by seberg** [comment]
Wait, did we never put some kind of warning/error about using this with weights in, sounds like this should be done still!? @madphysicist maybe you want to have a bit of a look at this, or @nayyarv has time to finish this off?

**Comment by madphysicist** [comment]
I skimmed this and the related (closed) PR. I will look into it in more detail as soon as I get a chance. The first thing that comes to mind is to add a `weightedfunctions` module similar to `nanfunctions`. However, even at my most optimistic, I have to admit this is *probably* not a reasonable solution.

**Comment by seberg** [comment]
Actually, I meant mostly if we are doing something reasonable with the
automatic bins right now. Error/warning ignore, I don't know, but just
to make sure it makes sense and is documented. We could try to make it
correct, but I expect it is not easy.

**Comment by njsmith** [comment]
@seberg: Yeah, I think you're right -- it looks like in master and 1.11b3, the new `bins="auto"` feature is kinda broken: `np.histogram` will happily accept `bins="auto", weights=...`, and then return nonsense results. Similarly for `bins="auto", range=...`. This PR looks like a good solution to me, but has conflicts due to other cleanups to this code...

@nayyarv @madphysicist: Any chance one of you could clean this up quickly so we can get it into 1.11?

**Comment by madphysicist** [comment]
I probably could some time this week, but certainly not over the next couple of days.

**Comment by madphysicist** [comment]
Cancel my previous comment. I took a look at the code more carefully. I can clean this up right now actually. Can I just rebase it onto #7199?

File: numpy/lib/function_base.py (hunk 5 above)
Comment (madphysicist) [inline numpy/lib/function_base.py]: How about a `RuntimeWarning`? Something along the lines of `FD estimator currently ignores weighted data`. Might be better than crashing.

**Comment by madphysicist** [comment]
I have rebased the `autobins` branch onto the latest `master` with all that that entails. I did add a couple of gramatical and line wrapping fixes. The result is PR #7243. I was unable to make a PR back to @nayyarv's fork of `numpy`.

@nayyarv, FWIW, I like the version in the `autobinsWeighted` branch better, even if this version is much simpler.

**Comment by madphysicist** [comment]
Looks like the other PR kicked off a rebuild on this one?

**Comment by nayyarv** [comment]
Nup, this was me trying to fix it, but looks like the rebase went a bit crazy.

**Comment by madphysicist** [comment]
Looks like you may have rebased master onto your branch instead of the other way around?

**Comment by madphysicist** [comment]
If you are OK with my PR, we can just run with that. I did not change any of the functionality, just some wording and whitespace stuff.

**Comment by nayyarv** [comment]
yep, that's exactly what happened. Let's jump over to your PR instead, I'm shutting this down.

**Comment by nayyarv** [comment]
Jump over to #7243, this pull request has gotten out of hand with my failure to git properly.

**Comment by charris** [comment]
You can recover from a failed rebase, `git reflog` is your friend, See http://stackoverflow.com/questions/134882/undoing-a-git-rebase.

**Comment by nayyarv** [comment]
Thanks charris, [I did manage to fix the rebase](https://github.com/numpy/numpy/compare/master...nayyarv:autobins) but since I closed this it, github won't let me reopen this Pull Request (since it's been recreated). As such, since madphysicist managed the exact same effect thing in his PR, I'm happy to let his go through instead

**Pull request closed without merging**